### ISSUE TITLE
`lax.custom_root` fails on root finders that require gradient

### ISSUE THREAD

**Issue by mbmccoy** [google/jax#8744]
The `lax.custom_root` method does not work on methods that require gradient, e.g., the Newton-Raphson method. The core of the problem seems to be overly-aggressive application of `stop_gradient` [here](https://github.com/google/jax/blob/e6950dd63ed3c054067c53bd4806dfbb85e0eef7/jax/_src/lax/control_flow.py#L2310).

While I've included a minimal example below, [this notebook](https://colab.research.google.com/drive/1e5_0gGYc-v9lKKwvUS7rqSUhA0D-OpT3) does a deeper dive into the error and demonstrates that removing `_stop_gradient_fun` applications resolves the problem.

I'd be happy to make a PR to resolve the issue, but my naïve solution of removing all calls to `stop_gradient` in `custom_root` appears... well, naïve. I'm just not sure why `stop_gradient` is required at all here, but it seems related to not wanting to trace gradients through control structures like while-loops. I suppose I'm curious why custom JVP'ing a function doesn't automatiically shortcut the gradient computation...

### Related issues

 - #1448 discusses a `define_implicit_gradient` function that doesn't use `stop_gradient`. This function was never merged. Regardless, the fact that `custom_root` methods can't use gradients seems like a bug.
   - Additional reading of the discussion there suggests that one of the main reasons for `stop_gradient` was the non-differentiability of control structures (`lax.while_loop` in particular). Nevertheless, my tests don't seem to trigger the issue.

## Minimal example

This example follows [the JAX unit test](https://github.com/google/jax/blob/e6950dd63ed3c054067c53bd4806dfbb85e0eef7/tests/lax_control_flow_test.py#L2010) of the `custom_root` method, with the bisection method replaced with Newton's method.

```python
import jax
import numpy as np

def newton(f, x0):
  """Newton's method for root-finding."""
  initial_state = (0, x0)  # (iteration, x)

  def cond(state):
    it, x = state
    # We fix 10 iterations for simplicity, this is plenty for convergence in our tests.
    return (it < 10)

  def body(state):
    it, x = state
    fx, dfx = f(x), jax.grad(f)(x)
    step = fx / dfx
    new_state = it + 1, x - step
    return new_state

  return jax.lax.while_loop(
    cond,
    body,
    initial_state,
  )[1]
```

### Newton's method without `custom_root`

```python
initial_guess = 2.0
x_test = 5.0

def sqrt_cubed_newton(x):
  implicit = lambda y: y ** 2 - x ** 3
  return newton(implicit, initial_guess)

print("Without custom_root")
v = sqrt_cubed_newton(x_test)
print(f"Value = {v}")
if abs(v - x_test ** 1.5) < 1e-5:
  print("Pass")
else:
  print("Fail")
```

```
Without custom_root
Value=11.180339813232422
Pass
```

### With `custom_root`

```python

def sqrt_cubed_newton_custom_root(x):
  def scalar_solve(g, y):
    return y / g(1.0)
  implicit = lambda y: y ** 2 - x ** 3
  return jax.lax.custom_root(implicit, initial_guess, newton, scalar_solve)

print("With custom_root")
v = sqrt_cubed_newton_custom_root(x_test)
print(f"Value = {v}")
if abs(v - x_test ** 1.5) < 1e-5:
  print("Pass")
else:
  print("Fail")
```
```
With custom_root
Value = nan
Fail
```

The notebook referenced above shows how removing the call to `_stop_gradient_fun` fixes the problem.

Please:

- [x] Check for duplicate issues.
- [x] Provide a complete example of how to reproduce the bug, wrapped in triple backticks like this
- [x] If applicable, include full error messages/tracebacks.

**Comment by shoyer** [comment]
Hi @mbmccoy, thanks for the careful investigation here!

I’m struggling to remember the exact context here, but I think we might actually be able to get away with removing the use of stop_gradient inside custom_root, because while_loop now has a JVP rule defined. This was not the case back when custom_root was written.

**Comment by shoyer** [comment]
We also switched from writing custom_root as a primitive to writing it via custom_jvp. So that also may have obviated the need for this.

Basically, I would suggest going ahead with removing stop_gradient_fun, being sure to add something like your Newton solver as an integration test.

**Comment by mbmccoy** [comment]
Thanks @shoyer for the comments.  I went ahead and removed it, and all tests continue to pass. I've added a few new tests for a newton solver in custom_root as well. 

PR here: https://github.com/google/jax/pull/8869
